File: src/net/bible/android/view/activity/bookmark/Bookmarks.java
package net.bible.android.view.activity.bookmark;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

import net.bible.android.activity.R;
import net.bible.android.control.ControlFactory;
import net.bible.android.control.bookmark.Bookmark;
import net.bible.android.view.activity.base.Callback;
import net.bible.android.view.activity.base.Dialogs;
import net.bible.android.view.activity.base.ListActivityBase;
import net.bible.android.view.activity.download.DownloadStatus;
import net.bible.android.view.activity.download.EnsureBibleDownloaded;
import net.bible.service.db.bookmark.BookmarkDto;
import net.bible.service.db.bookmark.LabelDto;
import net.bible.service.sword.SwordApi;

import org.crosswire.common.progress.JobManager;
import org.crosswire.common.util.Language;
import org.crosswire.jsword.book.Book;
import org.crosswire.jsword.book.BookFilter;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Spinner;
import android.widget.Toast;
import android.widget.AdapterView.OnItemSelectedListener;

/**
 * Choose Document (Book) to download
 * 
 * NotificationManager with ProgressBar example here:
 * http://united-coders.com/nico-heid/show-progressbar-in-notification-area-like-google-does-when-downloading-from-android
 * 
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br>
 *      The copyright to this program is held by it's author.
 */
public class Bookmarks extends ListActivityBase {
	private static final String TAG = "Bookmarks";

	private Bookmark bookmarkControl;
	
	// language spinner
	private Spinner labelSpinner;
	private List<LabelDto> labelList;
	private int selectedLabelNo = 0;
	private ArrayAdapter<LabelDto> labelArrayAdapter; 
	
	// the document list
	private ArrayAdapter<BookmarkDto> listArrayAdapter;
	private List<BookmarkDto> bookmarkList;

	private static final int LIST_ITEM_TYPE = R.layout.two_line_list_item_copy;

	private BookmarkDto selectedBookmark;
	
    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.bookmarks);

        bookmarkControl = ControlFactory.getInstance().getBookmarkControl();
        
       	initialiseView();
    }

    private void initialiseView() {
    	
    	bookmarkList = bookmarkControl.getAllBookmarks();
    	
    	// prepare the document list view
    	listArrayAdapter = new BookmarkItemAdapter(this, LIST_ITEM_TYPE, bookmarkList);
    	setListAdapter(listArrayAdapter);
    	
//    	//prepare the documentType spinner
//    	Spinner documentTypeSpinner = (Spinner)findViewById(R.id.documentTypeSpinner);
//    	documentTypeSpinner.setOnItemSelectedListener(new OnItemSelectedListener() {
//
//			@Override
//			public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
//		    	selectedDocumentFilterNo = position;
//		    	Bookmarks.this.filterDocuments();
//			}
//
//			@Override
//			public void onNothingSelected(AdapterView<?> arg0) {
//			}
//		});
//    	// in the basic flow we force the user to download a bible
//    	documentTypeSpinner.setEnabled(!forceBasicFlow);
//
//    	//prepare the language spinner
//    	{
//	    	langSpinner = (Spinner)findViewById(R.id.languageSpinner);
//	    	langSpinner.setOnItemSelectedListener(new OnItemSelectedListener() {
//	
//				@Override
//				public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
//			    	selectedLanguageNo = position;
//			    	Bookmarks.this.filterDocuments();
//				}
//	
//				@Override
//				public void onNothingSelected(AdapterView<?> arg0) {
//				}
//			});
//	    	langArrayAdapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, languageList);
//	    	langArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
//	    	langSpinner.setAdapter(langArrayAdapter);
//    	}
    }
    
    @Override
	protected void onListItemClick(ListView l, View v, int position, long id) {
//    	try {
//    		documentSelected(displayedDocuments.get(position));
//    	} catch (Exception e) {
//    		Log.e(TAG, "document selection error", e);
//    		showErrorMsg(R.string.error_occurred);
//    	}
	}

    private void populateMasterDocumentList() {
//		if (allDocuments==null || allDocuments.size()==0) {
//    	    new AsyncTask<Void, Boolean, Void>() {
//    	    	
//    	        @Override
//    	        protected void onPreExecute() {
//    	        	showHourglass();
//    	        }
//    	        
//    			@Override
//    	        protected Void doInBackground(Void... noparam) {
//    				try {
//	    	        	allDocuments = SwordApi.getInstance().getDownloadableDocuments();
//	    	        	for (Iterator<Book> iter=allDocuments.iterator(); iter.hasNext(); ) {
//	    	        		Book doc = iter.next();
//	    	        		if (doc.getLanguage()==null) {
//	    	        			Log.d(TAG, "Ignoring "+doc.getName()+" because it has no language");
//	    	        			iter.remove();
//	    	        		}
//	    	        		if (doc.getInitials().equals("WebstersDict")) {
//	    	        			Log.d(TAG, "Removing "+doc.getName()+" because it is too big and crashed dictionary code");
//	    	        			iter.remove();
//	    	        		}
//	    	        	}
//	    	        	Log.i(TAG, "number of documents available:"+allDocuments.size());
//    				} catch (Exception e) {
//    					Log.e(TAG, "Error getting documents to download", e);
//    					//todo INTERNATIONALIZE
//    					showErrorMsg("Error getting documents to download");
//    				}
//    	        	return null;
//    			}
//    			
//    	        @Override
//				protected void onPostExecute(Void result) {
//    	        	try {
//    	        		populateLanguageList();
//    	        		filterDocuments();
//    	        	} finally {
//    	        		//todo implement this: http://stackoverflow.com/questions/891451/android-dismissdialog-does-not-dismiss-the-dialog
//        	        	dismissHourglass();
//    	        	}
//    	        }
//
//    	    }.execute((Void[])null);
//		}
    }
    
    /** a spinner has changed so refilter the doc list
     */
    private void filterDocuments() {
//    	try {
//    		if (allDocuments!=null && allDocuments.size()>0) {
//   	        	Log.i(TAG, "filtering documents");
//	        	displayedDocuments.clear();
//	        	displayedDocumentDescriptions.clear();
//	        	String lang = languageList.get(selectedLanguageNo);
//	        	for (Book doc : allDocuments) {
//	        		BookFilter filter = DOCUMENT_TYPE_SPINNER_FILTERS[selectedDocumentFilterNo];
//	        		if (filter.test(doc) && doc.getLanguage().getName().equals(lang)) {
//		        		displayedDocuments.add(doc);
//		        		displayedDocumentDescriptions.add(doc.getName());
//	        		}
//	        	}
//	        	if (listArrayAdapter!=null) {
//	        		Bookmarks.this.listArrayAdapter.notifyDataSetChanged();
//	        	}
//    		}
//    	} catch (Exception e) {
//    		Log.e(TAG, "Error initialising view", e);
//    		Toast.makeText(this, getString(R.string.error)+e.getMessage(), Toast.LENGTH_SHORT);
//    	}
    }

    /** user selected a document so download it
     * 
     * @param document
     */
    private void documentSelected(Book document) {
//    	Log.d(TAG, "Document selected:"+document.getInitials());
//    	try {
//    		//sometimes the selected doc is null if the list was not clicked properly - odd!
//    		if (document!=null) {
//	    		this.selectedDocument = document;
//	
//	    		if (JobManager.getJobs().size()>=2) {
//	    			Log.i(TAG, "Too many jobs:"+JobManager.getJobs().size());
//	    			Dialogs.getInstance().showErrorMsg(R.string.too_many_jobs, new Callback() {
//						@Override
//						public void okay() {
//							showDownloadStatus();
//						}
//					});
//	    		} else {
//	    			showDialog(DOWNLOAD_CONFIRMATION_DIALOG);
//	    		}
//    		}
//    	} catch (Exception e) {
//    		Log.e(TAG, "Error on attempt to download", e);
//    		Toast.makeText(this, R.string.error_downloading, Toast.LENGTH_SHORT).show();
//    	}
    }

	@Override
    protected Dialog onCreateDialog(int id) {
//    	Dialog superDlg = super.onCreateDialog(id);
//    	if (superDlg!=null) {
//    		return superDlg;
//    	}
//    	
//        switch (id) {
//        case DOWNLOAD_CONFIRMATION_DIALOG:
//            	return new AlertDialog.Builder(this)
//            		   .setMessage(getText(R.string.download_document_confirm_prefix)+selectedDocument.getName())
//            	       .setCancelable(false)
//            	       .setPositiveButton(R.string.okay, new DialogInterface.OnClickListener() {
//            	           public void onClick(DialogInterface dialog, int id) {
//            	        	   doDownload(selectedDocument);
//            	           }
//            	       })
//            	       .setNegativeButton(R.string.cancel, new DialogInterface.OnClickListener() {
//            	           public void onClick(DialogInterface dialog, int id) {
//            	           }
//            	       }).create();
//        }
        return null;
    }

    @Override
	protected void onPrepareDialog(int id, Dialog dialog) {
//		super.onPrepareDialog(id, dialog);
//        switch (id) {
//        case DOWNLOAD_CONFIRMATION_DIALOG:
//        	AlertDialog alertDialog = (AlertDialog)dialog;
//        	alertDialog.setMessage(getText(R.string.download_document_confirm_prefix)+selectedDocument.getName());
//        };
	}

    
    @Override 
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
//    	if (resultCode==DOWNLOAD_FINISH) {
//    		returnToPreviousScreen();
//    	}
    }

//	@Override
//	protected void onStop() {
//		super.onStop();
//		langSpinner = null;
//		languageList = new ArrayList<String>();
//		langArrayAdapter = null;		
//		listArrayAdapter = null;
//		allDocuments = null;
//		displayedDocuments = null;
//		displayedDocumentDescriptions = null;
//		selectedDocument = null;
//	}
}

File: app/src/main/java/net/bible/android/view/activity/bookmark/Bookmarks.kt
/*
 * Copyright (c) 2020-2022 Martin Denham, Tuomas Airaksinen and the AndBible contributors.
 *
 * This file is part of AndBible: Bible Study (http://github.com/AndBible/and-bible).
 *
 * AndBible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * AndBible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with AndBible.
 * If not, see http://www.gnu.org/licenses/.
 */
package net.bible.android.view.activity.bookmark

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemLongClickListener
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.Toast
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import net.bible.android.activity.R
import net.bible.android.activity.databinding.BookmarksBinding
import net.bible.android.control.bookmark.BookmarkControl
import net.bible.android.control.event.ABEventBus
import net.bible.android.control.page.window.WindowControl
import net.bible.android.control.speak.SpeakControl
import net.bible.android.database.IdType
import net.bible.android.database.bookmarks.BookmarkEntities
import net.bible.android.view.activity.base.ListActionModeHelper
import net.bible.android.view.activity.base.ListActionModeHelper.ActionModeActivity
import net.bible.android.view.activity.base.ListActivityBase
import net.bible.service.common.CommonUtils.settings
import net.bible.android.database.bookmarks.BookmarkEntities.BaseBookmarkWithNotes
import net.bible.android.database.bookmarks.BookmarkEntities.Label
import net.bible.android.database.bookmarks.BookmarkSortOrder
import net.bible.android.view.activity.base.Dialogs
import net.bible.service.common.CommonUtils
import net.bible.service.common.displayName
import net.bible.service.db.BookmarksUpdatedViaSyncEvent
import java.lang.IllegalArgumentException
import java.util.*
import javax.inject.Inject

val BookmarkSortOrder.description get() =
    when(this) {
        BookmarkSortOrder.BIBLE_ORDER  -> CommonUtils.getResourceString(R.string.sort_by_bible_book)
        BookmarkSortOrder.LAST_UPDATED -> CommonUtils.getResourceString(R.string.sort_by_date)
        BookmarkSortOrder.CREATED_AT -> CommonUtils.getResourceString(R.string.sort_by_date)
        BookmarkSortOrder.CREATED_AT_DESC -> CommonUtils.getResourceString(R.string.sort_by_date)
        BookmarkSortOrder.ORDER_NUMBER -> "order number"
    }

/**
 * Choose Document (Book) to download
 *
 * NotificationManager with ProgressBar example here:
 * http://united-coders.com/nico-heid/show-progressbar-in-notification-area-like-google-does-when-downloading-from-android
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 */
class Bookmarks : ListActivityBase(), ActionModeActivity {

    private lateinit var binding: BookmarksBinding

    @Inject lateinit var bookmarkControl: BookmarkControl
    @Inject lateinit var speakControl: SpeakControl
    @Inject lateinit var windowControl: WindowControl

    private val labelList: MutableList<Label> = ArrayList()
    private var selectedLabelNo = 0

    // the document list
    private val bookmarkList: MutableList<BaseBookmarkWithNotes> = ArrayList()
    private var listActionModeHelper: ListActionModeHelper? = null
    override val integrateWithHistoryManager: Boolean = true

    /** Called when the activity is first created.  */
    @SuppressLint("MissingSuperCall")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ABEventBus.register(this)
        binding = BookmarksBinding.inflate(layoutInflater)
        setContentView(binding.root)
        settings.setLong("bookmarks-last-used", System.currentTimeMillis())
        buildActivityComponent().inject(this)

        // if coming Back using History then the LabelNo will be in the intent allowing the correct label to be pre-selected
        val extras = intent.extras
        if (extras != null) {
            if (extras.containsKey(BookmarkControl.LABEL_NO_EXTRA)) {
                val labelNo = extras.getInt(BookmarkControl.LABEL_NO_EXTRA)
                if (labelNo >= 0) {
                    selectedLabelNo = labelNo
                }
            }
        }
        initialiseView()
    }

    override fun onDestroy() {
        super.onDestroy()
        ABEventBus.unregister(this)
    }

    fun onEventMainThread(e: BookmarksUpdatedViaSyncEvent) {
        recreate()
    }

    private fun initialiseView() {
        listActionModeHelper = ListActionModeHelper(listView, R.menu.bookmark_context_menu)
        listView.onItemLongClickListener = OnItemLongClickListener{
            _, _, position, _ -> listActionModeHelper!!.startActionMode(this@Bookmarks, position)
        }

        //prepare the Label spinner
        loadLabelList()

        // check for pre-selected label e.g. when returning via History using Back button
        if (selectedLabelNo >= 0 && selectedLabelNo < labelList.size) {
            binding.labelSpinner.setSelection(selectedLabelNo)
        }
        binding.labelSpinner.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                selectedLabelNo = position
                loadBookmarkList()
            }

            override fun onNothingSelected(arg0: AdapterView<*>?) {}
        }

        // prepare the document list view
        val bookmarkArrayAdapter: ArrayAdapter<BaseBookmarkWithNotes> = BookmarkItemAdapter(
            this, bookmarkList, bookmarkControl, windowControl
        )
        listAdapter = bookmarkArrayAdapter
        loadBookmarkList()
    }

    override fun onListItemClick(l: ListView, v: View, position: Int, id: Long) {
        try {
            // check to see if Action Mode is in operation
            if (!listActionModeHelper!!.isInActionMode) {
                intent.putExtra("listPosition", position)
                bookmarkSelected(bookmarkList[position])
            }
        } catch (e: Exception) {
            Log.e(TAG, "document selection error", e)
            Dialogs.showErrorMsg(R.string.error_occurred, e)
        }
    }

    override val intentForHistoryList: Intent get()
    {
        Log.i(TAG, "Saving label no in History Intent")
        val intent = intent
        intent.putExtra(BookmarkControl.LABEL_NO_EXTRA, selectedLabelNo)
        return intent
    }

    private fun assignLabels(bookmarks: List<BaseBookmarkWithNotes>) = lifecycleScope.launch(Dispatchers.IO) {
        val labels = mutableSetOf<IdType>()
        for (b in bookmarks) {
            labels.addAll(bookmarkControl.labelsForBookmark(b).map { it.id })
        }

        val intent = Intent(this@Bookmarks, ManageLabels::class.java)
        intent.putExtra("data", ManageLabels.ManageLabelsData(
            mode = ManageLabels.Mode.ASSIGN,
            selectedLabels = labels
        ).applyFrom(windowControl.windowRepository.workspaceSettings).toJSON())
        val result = awaitIntent(intent)
        if(result.resultCode == RESULT_OK) {
            val resultData = ManageLabels.ManageLabelsData.fromJSON(result.data?.getStringExtra("data")!!)
            for (b in bookmarks) {
                bookmarkControl.changeLabelsForBookmark(b, resultData.selectedLabels.toList())
            }
            windowControl.windowRepository.workspaceSettings.updateFrom(resultData)
            withContext(Dispatchers.Main) {
                loadLabelList()
                loadBookmarkList()
            }
        }
    }

    private fun delete(bookmarks: List<BaseBookmarkWithNotes>) {
        AlertDialog.Builder(this)
            .setMessage(getString(R.string.confirm_delete_bookmarks, bookmarks.size))
            .setPositiveButton(R.string.yes) { _, _ ->
                for (bookmark in bookmarks) {
                    bookmarkControl.deleteBookmark(bookmark)
                }
                loadBookmarkList()
            }
            .setNegativeButton(R.string.cancel, null)
            .setCancelable(true)
            .show()
    }

    private fun loadLabelList() {
        labelList.clear()
        labelList.addAll(bookmarkControl.allLabels)
        val labelArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, labelList.map { it.displayName })
        labelArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        binding.labelSpinner.adapter = labelArrayAdapter
    }

    /** a spinner has changed so refilter the doc list
     */
    private fun loadBookmarkList() = lifecycleScope.launch(Dispatchers.IO) {
        withContext(Dispatchers.Main) {
            binding.empty.visibility = View.GONE
            binding.list.visibility = View.GONE
            binding.loadingIndicator.visibility = View.VISIBLE
        }
        try {
            if (selectedLabelNo > -1 && selectedLabelNo < labelList.size) {
                Log.i(TAG, "filtering bookmarks")
                val selectedLabel = labelList[selectedLabelNo]
                withContext(Dispatchers.Main) {
                    bookmarkList.clear()
                    bookmarkList.addAll(bookmarkControl.getBookmarksWithLabel(selectedLabel, bookmarkSortOrder))
                    bookmarkList.addAll(bookmarkControl.getGenericBookmarksWithLabel(selectedLabel))
                    notifyDataSetChanged()

                    // if in action mode then must exit because the data has changed, invalidating selections
                    listActionModeHelper!!.exitActionMode()
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error initialising view", e)
            withContext(Dispatchers.Main) {
                Toast.makeText(this@Bookmarks, getString(R.string.error) + " " + e.message, Toast.LENGTH_SHORT).show()
            }
        }
        withContext(Dispatchers.Main) {
            binding.loadingIndicator.visibility = View.GONE
            binding.list.visibility = View.VISIBLE
            binding.empty.visibility = View.VISIBLE
            listView.setSelection(intent.getIntExtra("listPosition", 0))
        }
    }

    private fun bookmarkSelected(bookmark: BaseBookmarkWithNotes) {
        Log.i(TAG, "Bookmark selected:$bookmark")
        try {
            if (bookmark is BookmarkEntities.BibleBookmarkWithNotes && bookmarkControl.isSpeakBookmark(bookmark)) {
                speakControl.speakFromBookmark(bookmark)
            }
            val resultIntent = Intent(this, Bookmarks::class.java)
            when(bookmark) {
                is BookmarkEntities.BibleBookmarkWithNotes -> {
                    resultIntent.putExtra("verse", bookmark.verseRange.start.osisID)
                }
                is BookmarkEntities.GenericBookmarkWithNotes -> {
                    resultIntent.putExtra("key", bookmark.key)
                    resultIntent.putExtra("book", bookmark.book?.initials)
                    resultIntent.putExtra("ordinal", bookmark.ordinalStart)
                }
            }
            setResult(Activity.RESULT_OK, resultIntent)
            finish()
        } catch (e: Exception) {
            Log.e(TAG, "Error on bookmarkSelected", e)
            Toast.makeText(this, R.string.error_occurred, Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        val inflater = menuInflater
        inflater.inflate(R.menu.bookmark_actionbar_menu, menu)
        return true
    }

    private fun changeBookmarkSortOrder() {
        bookmarkSortOrder = when (bookmarkSortOrder) {
            BookmarkSortOrder.BIBLE_ORDER -> BookmarkSortOrder.CREATED_AT_DESC
            BookmarkSortOrder.CREATED_AT_DESC -> BookmarkSortOrder.CREATED_AT            
            BookmarkSortOrder.CREATED_AT -> BookmarkSortOrder.BIBLE_ORDER
            else -> BookmarkSortOrder.CREATED_AT_DESC
        }
    }

    private var bookmarkSortOrder: BookmarkSortOrder
        get() {
            val bookmarkSortOrderStr = CommonUtils.getSharedPreference(BOOKMARK_SORT_ORDER, BookmarkSortOrder.BIBLE_ORDER.toString())
            return try {
                BookmarkSortOrder.valueOf(bookmarkSortOrderStr!!)
            } catch (e: IllegalArgumentException) { BookmarkSortOrder.BIBLE_ORDER }
        }
        private set(bookmarkSortOrder) {
            CommonUtils.saveSharedPreference(BOOKMARK_SORT_ORDER, bookmarkSortOrder.toString())
        }

    /**
     * on Click handlers
     */
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        var isHandled = false
        when (item.itemId) {
            R.id.sortByToggle -> {
                isHandled = true
                try {
                    changeBookmarkSortOrder()
                    val sortDesc = bookmarkSortOrder.description
                    Toast.makeText(this, sortDesc, Toast.LENGTH_SHORT).show()
                    loadBookmarkList()
                } catch (e: Exception) {
                    Log.e(TAG, "Error sorting bookmarks", e)
                    Dialogs.showErrorMsg(R.string.error_occurred, e)
                }
            }
            R.id.manageLabels -> {
                isHandled = true
                lifecycleScope.launch(Dispatchers.Main) {
                    val intent = Intent(this@Bookmarks, ManageLabels::class.java)
                    intent.putExtra("data", ManageLabels.ManageLabelsData(
                        mode = ManageLabels.Mode.WORKSPACE,
                    ).applyFrom(windowControl.windowRepository.workspaceSettings).toJSON())
                    val result = awaitIntent(intent)
                    if(result.resultCode == RESULT_OK) {
                        val resultData = ManageLabels.ManageLabelsData.fromJSON(result.data?.getStringExtra("data")!!)
                        windowControl.windowRepository.workspaceSettings.updateFrom(resultData)
                        withContext(Dispatchers.Main) {
                            loadLabelList()
                            loadBookmarkList()
                        }
                    }
                }
            }
        }
        if (!isHandled) {
            isHandled = super.onOptionsItemSelected(item)
        }
        return isHandled
    }

    override fun onActionItemClicked(item: MenuItem, selectedItemPositions: List<Int>): Boolean {
        val selectedBookmarks = getSelectedBookmarks(selectedItemPositions)
        when (item.itemId) {
            R.id.assign_labels -> assignLabels(selectedBookmarks)
            R.id.delete -> delete(selectedBookmarks)
        }
        return true
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if(requestCode == REQUEST_MANAGE_LABELS) {
            loadLabelList()
            loadBookmarkList()
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    override fun isItemChecked(position: Int): Boolean {
        return listView.isItemChecked(position)
    }

    private fun getSelectedBookmarks(selectedItemPositions: List<Int>): List<BaseBookmarkWithNotes> {
        val selectedBookmarks: MutableList<BaseBookmarkWithNotes> = ArrayList()
        for (position in selectedItemPositions) {
            selectedBookmarks.add(bookmarkList[position])
        }
        return selectedBookmarks
    }

    companion object {
        private const val BOOKMARK_SORT_ORDER = "BookmarkSortOrder"
        private const val TAG = "Bookmarks"
        private const val REQUEST_MANAGE_LABELS = 10
    }
}
